### PR TITLE
Add support for error field to createAction

### PR DESCRIPTION
**Pull request by sjsyrek**
Add support for an `error` field to `createAction`. This change does _not_ make it a requirement that the `payload` is an `Error` object, and the `error` field itself is optional, both of these in accordance with the [Flux standard action](https://github.com/redux-utilities/flux-standard-action#actions) description. The `error` is also not required to be boolean, as it seems up to the client to determine what to do with these values.

Thanks for a great library! I really want to use this API to reduce boilerplate, but I need the `error` field, so I thought I'd offer the minimal change required to get that functioning. I added some tests and updated the README as best I could (including some general grammatical fixes) to reflect the changes.

_Usage_
```ts
const action = createAction(types.WITH_PAYLOAD_META_ERROR, resolve => {
  return (id: number, token: string, error: boolean) => resolve(id, token, error);
});

const actual: {
  type: 'WITH_PAYLOAD_META_ERROR';
  payload: number;
  meta: string;
  error: boolean;
} = action(1, 'token', false); // <- it makes me happy to add `false` here :)
```

_Notes_
- I couldn't find a `check` script, so I assume that meant `npm run ci-check`
- I started to add "type unit tests" assuming that meant `types.spec.ts`, but these depend on utils
that use with `createStandardAction`, and I didn't want to start making my modification larger just for this purpose
- I couldn't find the "runtime unit tests"

_Related Issue_
#52 

* [x] Rebase before creating a PR to keep commit history clean
* [x] Clear node_modules and reinstall all the dependencies: `npm run reinstall`
* [x] Run checker npm script: `npm run check`

Extra checklist:
* [x] Always add some description and refer to all the related issues using `#` tag

For bugfixes:
 * [x] Add at least one unit test to cover the bug that have been fixed

For new feature:
 * [x] Update API docs
 * [x] Add examples to demonstrate new feature
 * [ ] Add type unit tests
 * [ ] Add runtime unit tests

### CODE REVIEW

**Comment by sjsyrek** [comment]
Hi @piotrwitek do you have any feedback for this PR? Just checking in, as I see the merge conflicts accumulate :-)

File: src/create-action.ts [hunk 1]
```diff
@@ -2,13 +2,31 @@ import { StringType, ActionCreator } from './types';
...
-    ? { type: T }
-    : { type: T; meta: M }
-  : M extends undefined
-  ? { type: T; payload: P }
-  : { type: T; payload: P; meta: M };
+type ActionWithType<T> = { type: T };
+type ActionWithPayload<T, P> = { type: T; payload: P };
+type ActionWithMeta<T, M> = { type: T; meta: M };
+type ActionWithPayloadAndMeta<T, P, M> = { type: T; payload: P; meta: M };
+type ActionWithError<T, E> = { type: T; error: E };
```
Comment (piotrwitek) [inline src/create-action.ts]: ActionWithError type is the same as ActionWithMeta, without constrains it will not work

**Comment by sjsyrek** [comment]
@piotrwitek Thanks for the feedback, this is great. I will probably have to do some trial and error plus research, but at least I know it's worth working on 👍

**Comment by piotrwitek** [comment]
@sjsyrek you're welcome :)

I reminded myself one thing that I was experimenting with something similar in the past and I hit the wall that it was impossible to do with current type system limitations, especially stay away from overloads (overloads does break in few type inference scenarios), conditionals are the way to go. That's why I have refactored a lot of internal types from overloads to conditional and I still have some of the refactoring left on a local branch that I plan to merge this week. So I would wait for this because then you'll have a cleaner and more robust baseline to work on.

**Comment by sjsyrek** [comment]
@piotrwitek I saw the recent changes... do they make this PR unnecessary? Or is it still useful to have a FSA creator that includes an error field? I like the brevity of `createAction`.

**Comment by piotrwitek** [comment]
Hey @sjsyrek, yes I would definitely extend createAction to cover this common case, but I would do it with slightly different API.
Your proposal is IMO too limited, covering only a single use-case. Imagine in a few days somebody else would like to add another property and we're stuck with bad API. Also note there is `createStandardAction` that is FSA compliant so if you are trying to chase FSA then I would suggest to look there.

As for solution proposal I would go with a third argument being a bag object, that will be merged with action object. Here is a usage example:
```ts
createAction(types.WITH_PAYLOAD_META_ERROR, action => {
  return (id: number, token: string, error: boolean) => action(id, token, { error });
});
```

I would accept this solution as this will cover this and many different cases in the future.

If you're interested in tackling it, please copy this proposal and create a new issue, where we will continue API design discussion.

**Comment by sjsyrek** [comment]
@piotrwitek Sure, I could take a crack at it. Created the issue here: https://github.com/piotrwitek/typesafe-actions/issues/116

Side note: would it not also be worth having support for Redux Thunk, so you could also use `ActionType` with thunk actions? Since `redux-thunk` comes with a `ThunkAction` type, it might not be too difficult to add in (separate issue).

**Comment by piotrwitek** [comment]
@sjsyrek as for `redux-thunk` if that is possible then I think it's a good idea. But note that I don't want to include a third-party dependency, so the idea would be to add an internal common interface with redux-thunk action type. Also we could post a working solution in the documentation.

**Comment by piotrwitek** [comment]
@sjsyrek I think this is obsolete right? Should I close it?

**Comment by sjsyrek** [comment]
@piotrwitek I haven't had time recently to look at this but would still like to contribute if there's a need. If you think this is obsolete, I'm fine with closing it.

**Comment by piotrwitek** [comment]
Hey @sjsyrek, thanks a lot for stepping in!
I just returned to work after the summer break, I'll be reviewing everything in the coming days, I will notify you later.

**Comment by piotrwitek** [comment]
Hey @sjsyrek,
I would like to get back to work on this feature in v5.x.x.

1) How it would work type-wise?
Most probably and optional error field added to base Action type:
```ts
type Action = {
  type: string;
  error?: boolean;
}
```

2) How it would work at runtime?
I would go with a forced convention by having a payload to be an instance of Error (easy to implement, backward compatible does not change API and I think that's the most popular option) 
This would automatically set error property to true, but also we would have a chance to use conditional types to refine Action type to show or hide error property from action type. That would be the most optimal solution.

What are your thoughts?

PS: I would like to close this PR because it's outdated and create a new issue tracking this feature.

**Comment by piotrwitek** [comment]
Closing this in favor of a new proposal at #207

**Pull request closed without merging**